README.md
<a href="http://www.boost.org/LICENSE_1_0.txt" target="_blank">![Boost Licence](http://img.shields.io/badge/license-boost-blue.svg)</a>
<a href="https://github.com/cpp-testing/GUnit/releases" target="_blank">![Version](https://badge.fury.io/gh/cpp-testing%2FGUnit.svg)</a>
<a href="https://travis-ci.org/cpp-testing/GUnit" target="_blank">![Build Status](https://img.shields.io/travis/cpp-testing/GUnit/master.svg?label=linux/osx)</a>
<a href="https://codecov.io/gh/cpp-testing/GUnit" target="_blank">![Coveralls](https://codecov.io/gh/cpp-testing/GUnit/branch/master/graph/badge.svg)</a>
<a href="http://github.com/cpp-testing/GUnit/issues" target="_blank">![Github Issues](https://img.shields.io/github/issues/cpp-testing/GUnit.svg)</a>

---

### Testing

> "If you liked it then you should have put a test on it", Beyonce rule

## GUnit
> [Google.Test/Google.Mock](https://github.com/google/googletest)/[Cucumber](https://github.com/cucumber/cucumber/wiki/Gherkin) on steroids

* Improve your productivity with GUnit, a library which extends/simplifies [Google.Test/Google.Mock](https://github.com/google/googletest) and adds support for [Gherkin](https://github.com/cucumber/cucumber/wiki/Gherkin) (Behaviour Driven Development) to it.

    * Why it's based on [Google.Test/Google.Mock](https://github.com/google/googletest)?
        * (+) Google.Test is **widely used** (The most popular testing framework according to https://www.jetbrains.com/research/devecosystem-2017/cpp)
        * (+) Google.Test is **stable**
        * (+) Google.Test is **powerful**
        * (+) Google.Test comes with **Google.Mock**
        * (+) Google.Test is **well documented**
        * (-) Google.Test **doesn't have support for - [gherkin](https://github.com/cucumber/cucumber/wiki/Gherkin) style - tests**
        * (-) Google.Test and Google.Mock have a lot **boilerplate macros**

### Motivation Examples

> #### No more base classes and labels as identifiers - [GUnit.GTest](docs/GTest.md) / [GUnit.GTest-Lite](docs/GTest-Lite.md)
  ```cpp
                 Google.Test                        |                     GUnit.GTest
  --------------------------------------------------+------------------------------------------------------
  #include <gtest/gtest.h>                          | #include <Gunit.h>
                                                    |
  struct CalcTest : testing::Test {                 | GTEST("Calc Test") {
   void SetUp() override {                          |   Calc calc{};
     calc = std::make_unique<Calc>();               |
   }                                                |   // SetUp
                                                    |
   void TearDown() override { }                     |   SHOULD("return sum of 2 numbers") {
                                                    |     EXPECT_EQ(5, calc->add(4, 1));
   std::unique_ptr<Calc> calc;                      |   }
  };                                                |
                                                    |   SHOULD("throw if division by 0") {
  TEST_F(ExampleTest, ShouldReturnSumOf2Numbers) {  |     EXPECT_ANY_THROW(calc->div(42, 0));
    EXPECT_EQ(5, calc->add(4, 1));                  |   }
  }                                                 |
                                                    |   // TearDown
  TEST_F(ExampleTest, ShouldThrowIfDivisionBy0) {   | }
    EXPECT_ANY_THROW(calc->div(42, 0));             |
  }                                                 |
  ```

  > Output
  ```sh
  [----------] 2 tests from CalcTest                | [----------] 1 tests from Calc Test
  [ RUN      ] CalcTest.ShouldReturnSumOf2Numbers   | [ RUN      ] Calc Test
  [       OK ] CalcTest.ShouldReturnSumOf2Numbers   | [ SHOULD   ] return sum of 2 numbers
  [ RUN      ] CalcTest.ShouldThrowIfDivisionBy0    | [ SHOULD   ] throw if division by 0
  [       OK ] CalcTest.ShouldThrowIfDivisionBy0    | [       OK ] Calc Test (0 ms)
  [----------] 2 tests from CalcTest (1 ms total)   | [----------] 1 tests from Example (0 ms total)
  ```

> #### No more hand written mocks - [GUnit.GMock](docs/GMock.md)
  ```cpp
  struct interface {
    virtual ~interface() = default;
    virtual int get() const = 0;
    virtual void foo(int) = 0;
    virtual void bar(int, const std::string&) = 0;
  };
  ```
  ```cpp
                 Google.Test                        |                     GUnit.GMock
  --------------------------------------------------+------------------------------------------------------
  #include <gmock/gmock.h>                          | #include <Gunit.h>
                                                    |
  struct mock_interface : interface {               |
    MOCK_CONST_METHOD0(get, int(int));              |
    MOCK_METHOD1(foo, void(int));                   |
    MOCK_METHOD2(bar, void(int, const string&));    |
  };                                                |
                                                    |
  int main() {                                      | int main() {
    StrictMock<mock_interface> mock{};              |   StrictGMock<interface> mock{};
    EXPECT_CALL(mock, foo(42));                     |   EXPECT_CALL(mock, (foo)(42));
                                                    |
    interface& i = mock;                            |   interface& i = mock.object();
    i.foo(42);                                      |   i.foo(42);
  }                                                 | }
  ```

> #### Simplified creation and injection of SUT (System Under Test) and mocks - [GUnit.GMake](docs/GMake.md)
  ```cpp
  class coffee_maker {
   public:
     coffee_maker(iheater&, ipump&, igrinder&);
     ...
  };
  ```
  ```cpp
                 Google.Test                        |                     GUnit.GMake
  --------------------------------------------------+------------------------------------------------------
   #include <gtest/gtest.h>                         | #include <Gunit.h>
   #include <gmock/gmock.h>                         |
                                                    |
   TEST(CalcTest, ShouldMakeCoffee) {               | GTEST("Calc Test") {
     StrictMock<mock_heater> heater{};              |   std::tie(sut, mocks) = // auto [sut, mocks] in C++17
     StrictMock<mock_pump> pump{};                  |     make<coffee_maker, StrictGMock>();
     StrictMock<mock_grinder> grinder{};            |
     coffee_maker sut{heater, pump, grinder};       |   EXPECT_CALL(mocks.mock<iheater>(), (on)());
                                                    |   EXPECT_CALL(mocks.mock<ipump>(), (pump)());
     EXPECT_CALL(heater, on());                     |   EXPECT_CALL(mocks.mock<igrinder>(), (grind)());
     EXPECT_CALL(pump, pump());                     |   EXPECT_CALL(mocks.mock<iheater>(), (off)());
     EXPECT_CALL(grinder, grind());                 |
     EXPECT_CALL(heater, off());                    |   sut->brew();
                                                    | }
     sut->brew();                                   |
   }
  ```

> #### Support for - [Gherkin](https://github.com/cucumber/cucumber/wiki/Gherkin) style - BDD (Behaviour Driven Development) scenarios - [GUnit.GSteps](docs/GSteps.md)

  > Feature specification

    Test/Features/Calc/addition.feature

  ```gherkin
  Feature: Calc Addition
    In order to avoid silly mistakes
    As a math idiot
    I want to be told the sum of two numbers

    Scenario: Add two numbers
      Given I created a calculator with value 0
        And I have entered 20 into the calculator
        And I have entered 30 into the calculator
       When I press add
       Then The result should be 50
  ```

  > Steps Implementation

    Test/Features/Calc/Steps/CalcSteps.cpp

  ```cpp
  #include <Gunit.h>

  GSTEPS("Calc*") { // "Calc Addition.Add two numbers"
    auto result = 0;

    Given("I created a calculator with value {n}") = [&](int n) {
      Calculator calc{n};

      Given("I have entered {n} into the calculator") = [&](int n) {
        calc.push(n);
      };

      When("I press add") = [&] {
        result = calc.add();
      };

      Then("The result should be {expected}") = [&](int expected) {
         EXPECT_EQ(expected, result);
      };
    };
  }
  ```

  > Usage
  ```sh
  SCENARIO="Test/Features/Calc/addition.feature" ./test --gtest_filter="Calc Addition.Add two numbers"
  ```

  > Output
  ```sh
  [==========] Running 1 test from 1 test case.
  [----------] Global test environment set-up.
  [----------] 1 tests from Calc Addition
  [ RUN      ] Calc Addition.Add two numbers
  [    Given ] I have created a calculator with value 0         # CalcSteps.cpp:10
  [    Given ] I have entered 20 into the calculator            # CalcSteps.cpp:12
  [    Given ] I have entered 30 into the calculator            # CalcSteps.cpp:14
  [     When ] I press add                                      # CalcSteps.cpp:16
  [     Then ] the result should be 50 on the screen            # CalcSteps.cpp:19
  [----------] Global test environment tear-down
  [==========] 1 test from 1 test case ran. (7 ms total)
  [  PASSED  ] 1 tests.
  ```

### Overview

* `GUnit.GTest` - Google.Test with strings and more friendly macros
  * Test cases with string as names
  * No more SetUp/TearDown (SHOULD clauses)
  * One (GTEST) macro for all types of tests
  * 100% Compatible with tests using GTest
* `GUnit.GTest-Lite` - lightweight, limited, no-macro way of defining simple tests
* `GUnit.GMock` - Google.Mock without hand written mocks
  * No more hand written mocks!
  * Support for more than 10 parameters
  * Quicker compilation times
  * Support for unique_ptr without any tricks
  * Support for overloaded operators
  * Support for mocking classes with constructors
  * 100% Compatible with Google Mocks
* `GUnit.GMake` - Makes creation of System Under Test (SUT) and Mocks easier
  * No need to instantiate SUT (System Under Test) and mocks
    * Automatic mocks injection
* `GUnit.GSteps` - Behaviour Driven Development
  * Support for - [Gherkin](https://github.com/cucumber/cucumber/wiki/Gherkin) style - BDD tests

### Quick Start

* If your project is **NOT** using [Google.Test/Google.Mock](https://github.com/google/googletest)
    * Follow instructions from https://github.com/google/googletest/tree/master/googletest
* Clone the repository
    * `git clone https://github.com/cpp-testing/GUnit.git`
* Add `GUnit/include` directory to your include path
    * `-I GUnit/include`
* Write some tests...
* Compile and Run!

---

* [**Optional**] For [gherkin](https://github.com/cucumber/cucumber/wiki/Gherkin) support
    * Compile `gherkin-cpp`
    ```sh
    $cd libs/gherkin-cpp && make lib
    $ls libs/gherkin-cpp
      libgherkin-cpp.a
      libgherkin-cpp.so
    ```
    * Add include paths
        * `-I GUnit/gherkin-cpp/include`
        * `-I GUnit/json/src`
    * Link with `libgherkin-cpp.{a, so}`
        * `-L libgherkin-cpp`
    * Write some feature tests...
    * Compile and Run!

---

* To run GUnit tests/benchmarks
  ```sh
  $mkdir build && cd build && cmake ..
  $make && ctest
  ```

### Requirements
  * [C++14](https://ubershmekel.github.io/cppdrafts/c++14-cd.html)
  * `GTest/GTest-Lite/GMock/GMake`
    * [libs/googletest](https://github.com/google/googletest) - compatible with all versions
  * `GSteps`
    * [libs/json](https://github.com/nlohmann/json)
    * [libs/gherkin-cpp](https://github.com/c-libs/gherkin-cpp)

### Tested compilers
  * [Linux - GCC-5+](https://travis-ci.org/cpp-testing/GUnit)
  * [Linux - Clang-3.7+](https://travis-ci.org/cpp-testing/GUnit)
  * [Mac OSx - Xcode-7.3+](https://travis-ci.org/cpp-testing/GUnit)

### User Guide
  * **[GUnit.GTest](docs/GTest.md)**
  * **[GUnit.GTest-Lite](docs/GTest-Lite.md)**
  * **[GUnit.GMock](docs/GMock.md)**
  * **[GUnit.GMake](docs/GMake.md)**
  * **[GUnit.GSteps](docs/GSteps.md)**

### [FAQ](docs/FAQ.md)
  * [C++ Now 2017: Towards Painless Testing](https://www.youtube.com/watch?v=NVrZjT5lW5o)

### Acknowledgements
* Thanks to [Google Open Source](https://opensource.google.com/) for [Google.Test/Google.Mock](https://github.com/google/googletest) libraries
* Thanks to Eran Pe'er for [FakeIt](https://github.com/eranpeer/FakeIt) library
* Thanks to Peter Bindels for [HippoMocks](https://github.com/dascandy/hippomocks) library
* Thanks to Niels Lohmann for [json](https://github.com/nlohmann/json) library
* Thanks to Aslak Hellesøy for [gherkin-c](https://github.com/cucumber/gherkin-c) library
* Thanks to [Cucumber Open Source](https://cucumber.io/) for [cucumber-cpp](https://github.com/cucumber/cucumber-cpp) library


CHANGELOG.md
## [1.0.0] - 2017-12-15
- Support for Linux - Clang-3.7+
- Support for Linux - GCC-5+
- Support for Mac OSx - Xcode-7.3+

---

[1.0.0]: https://github.com/cpp-testing/GUnit/tree/v1.0.0


include/gmock.vptr.h
//
// Copyright (c) 2016 Krzysztof Jusiak (krzysztof at jusiak dot net)
//
// Distributed under the Boost Software License, Version 1.0.
// (See accompanying file LICENSE_1_0.txt or copy at
// http://www.boost.org/LICENSE_1_0.txt)
//
#pragma once

#include <gmock/gmock.h>
#include <gtest/gtest.h>  //TODO split into 2 files
//#include <libunwind.h>
#include <memory>
#include <string>
#include <tuple>
#include <type_traits>
#include <unordered_map>

#include <cxxabi.h>
#include <execinfo.h>
#include <stdio.h>
#include <stdlib.h>

#define __REQUIRES(...) typename std::enable_if<__VA_ARGS__, int>::type

/** Print a demangled stack backtrace of the caller function to FILE* out. */
static inline void print_stacktrace(FILE* out = stderr, unsigned int max_frames = 63) {
  fprintf(out, "stack trace:\n");

  // storage array for stack trace address data
  void* addrlist = new void*[max_frames + 1];

  // retrieve current stack addresses
  int addrlen = backtrace((void**)addrlist, sizeof(addrlist) / sizeof(void*));

  if (addrlen == 0) {
    fprintf(out, "  <empty, possibly corrupt>\n");
    return;
  }

  // resolve addresses into strings containing "filename(function+address)",
  // this array must be free()-ed
  char** symbollist = backtrace_symbols((void* const*)addrlist, addrlen);

  // allocate string which will be filled with the demangled function name
  size_t funcnamesize = 256;
  char* funcname = (char*)malloc(funcnamesize);

  // iterate over the returned symbol lines. skip the first, it is the
  // address of this function.
  for (int i = 1; i < addrlen; i++) {
    char *begin_name = 0, *begin_offset = 0, *end_offset = 0;

    // find parentheses and +address offset surrounding the mangled name:
    // ./module(function+0x15c) [0x8048a6d]
    for (char* p = symbollist[i]; *p; ++p) {
      if (*p == '(')
        begin_name = p;
      else if (*p == '+')
        begin_offset = p;
      else if (*p == ')' && begin_offset) {
        end_offset = p;
        break;
      }
    }

    if (begin_name && begin_offset && end_offset && begin_name < begin_offset) {
      *begin_name++ = '\0';
      *begin_offset++ = '\0';
      *end_offset = '\0';

      // mangled name is now in [begin_name, begin_offset) and caller
      // offset in [begin_offset, end_offset). now apply
      // __cxa_demangle():

      int status;
      char* ret = abi::__cxa_demangle(begin_name, funcname, &funcnamesize, &status);
      if (status == 0) {
        funcname = ret;  // use possibly realloc()-ed string
        fprintf(out, "  %s : %s+%s\n", symbollist[i], funcname, begin_offset);
      } else {
        // demangling failed. Output function name as a C function with
        // no arguments.
        fprintf(out, "  %s : %s()+%s\n", symbollist[i], begin_name, begin_offset);
      }
    } else {
      // couldn't parse the line? print the whole line.
      fprintf(out, "  %s\n", symbollist[i]);
    }
  }

  free(funcname);
  free(symbollist);
}

namespace testing {
inline namespace v1 {
template <class>
class GMock;

namespace detail {

template <typename T>
struct type {
  static void id() {}
};

template <typename T>
std::size_t type_id() {
  return reinterpret_cast<std::size_t>(&type<T>::id);
}

template <class T, class... TArgs>
decltype(void(T{std::declval<TArgs>()...}), std::true_type{}) test_is_braces_constructible(int);
template <class, class...>
std::false_type test_is_braces_constructible(...);
template <class T, class... TArgs>
using is_braces_constructible = decltype(test_is_braces_constructible<T, TArgs...>(0));
template <class T, class... TArgs>
using is_braces_constructible_t = typename is_braces_constructible<T, TArgs...>::type;

template <class TParent>
struct any_type {
  template <class T, __REQUIRES(std::is_polymorphic<std::decay_t<T>>::value) = 0>
  operator T&() {
    auto mock = std::make_shared<GMock<std::decay_t<T>>>();
    mocks[type_id<std::decay_t<T>>()] = mock;
    return *mock;
  }

  template <class T, __REQUIRES(!std::is_polymorphic<T>::value) = 0>
  operator T() {
    return {};
  }

  std::unordered_map<std::size_t, std::shared_ptr<void>>& mocks;
};

template <class T>
auto make_impl() {
  std::unordered_map<std::size_t, std::shared_ptr<void>> mocks;
  return std::make_pair(std::make_unique<T>(any_type<T>{mocks}, any_type<T>{mocks}), mocks);
}

template <class T>
struct identity {
  using type = T;
};

template <class T>
using identity_t = typename identity<T>::type;

template <class TDst, class TSrc>
inline TDst union_cast(TSrc src) {
  union {
    TSrc src;
    TDst dst;
  } u;
  u.src = src;
  return u.dst;
}

template <char... Chrs>
struct string {
  static auto c_str() {
    static constexpr char str[] = {Chrs..., 0};
    return str;
  }
};

template <class T>
inline auto vptr_offset(T f) {
  struct vptr {
    virtual int f0() { return 0; }
    virtual int f1() { return 1; }
    virtual int f2() { return 2; }
    virtual int f3() { return 3; }
    virtual int f4() { return 4; }
    virtual int f5() { return 5; }
    virtual int f6() { return 6; }
    virtual int f7() { return 7; }
    virtual int f8() { return 8; }
    virtual int f9() { return 9; }
    virtual int f10() { return 10; }
    virtual int f11() { return 11; }
    virtual int f12() { return 12; }
    virtual int f13() { return 13; }
    virtual int f14() { return 14; }
    virtual int f15() { return 15; }
    virtual int f16() { return 16; }
    virtual int f17() { return 17; }
    virtual int f18() { return 18; }
    virtual int f19() { return 19; }
    virtual int f20() { return 20; }
    virtual int f21() { return 21; }
    virtual int f22() { return 22; }
    virtual int f23() { return 23; }
    virtual int f24() { return 24; }
    virtual int f25() { return 25; }
    virtual int f26() { return 26; }
    virtual int f27() { return 27; }
    virtual int f28() { return 28; }
    virtual int f29() { return 29; }
    virtual int f30() { return 30; }
    virtual int f31() { return 31; }
    virtual int f32() { return 32; }
    virtual int f33() { return 33; }
    virtual int f34() { return 34; }
    virtual int f35() { return 35; }
    virtual int f36() { return 36; }
    virtual int f37() { return 37; }
    virtual int f38() { return 38; }
    virtual int f39() { return 39; }
    virtual int f40() { return 40; }
    virtual int f41() { return 41; }
    virtual int f42() { return 42; }
    virtual int f43() { return 43; }
    virtual int f44() { return 44; }
    virtual int f45() { return 45; }
    virtual int f46() { return 46; }
    virtual int f47() { return 47; }
    virtual int f48() { return 48; }
    virtual int f49() { return 49; }
    virtual int f50() { return 50; }
    virtual int f51() { return 51; }
    virtual int f52() { return 52; }
    virtual int f53() { return 53; }
    virtual int f54() { return 54; }
    virtual int f55() { return 55; }
    virtual int f56() { return 56; }
    virtual int f57() { return 57; }
    virtual int f58() { return 58; }
    virtual int f59() { return 59; }
    virtual int f60() { return 60; }
    virtual int f61() { return 61; }
    virtual int f62() { return 62; }
    virtual int f63() { return 63; }
    virtual int f64() { return 64; }
    virtual int f65() { return 65; }
    virtual int f66() { return 66; }
    virtual int f67() { return 67; }
    virtual int f68() { return 68; }
    virtual int f69() { return 69; }
    virtual int f70() { return 70; }
    virtual int f71() { return 71; }
    virtual int f72() { return 72; }
    virtual int f73() { return 73; }
    virtual int f74() { return 74; }
    virtual int f75() { return 75; }
    virtual int f76() { return 76; }
    virtual int f77() { return 77; }
    virtual int f78() { return 78; }
    virtual int f79() { return 79; }
    virtual int f80() { return 80; }
    virtual int f81() { return 81; }
    virtual int f82() { return 82; }
    virtual int f83() { return 83; }
    virtual int f84() { return 84; }
    virtual int f85() { return 85; }
    virtual int f86() { return 86; }
    virtual int f87() { return 87; }
    virtual int f88() { return 88; }
    virtual int f89() { return 89; }
    virtual int f90() { return 90; }
    virtual int f91() { return 91; }
    virtual int f92() { return 92; }
    virtual int f93() { return 93; }
    virtual int f94() { return 94; }
    virtual int f95() { return 95; }
    virtual int f96() { return 96; }
    virtual int f97() { return 97; }
    virtual int f98() { return 98; }
    virtual int f99() { return 99; }
    virtual int f100() { return 100; }
    virtual int f101() { return 101; }
    virtual int f102() { return 102; }
    virtual int f103() { return 103; }
    virtual int f104() { return 104; }
    virtual int f105() { return 105; }
    virtual int f106() { return 106; }
    virtual int f107() { return 107; }
    virtual int f108() { return 108; }
    virtual int f109() { return 109; }
    virtual int f110() { return 110; }
    virtual int f111() { return 111; }
    virtual int f112() { return 112; }
    virtual int f113() { return 113; }
    virtual int f114() { return 114; }
    virtual int f115() { return 115; }
    virtual int f116() { return 116; }
    virtual int f117() { return 117; }
    virtual int f118() { return 118; }
    virtual int f119() { return 119; }
    virtual int f120() { return 120; }
    virtual int f121() { return 121; }
    virtual int f122() { return 122; }
    virtual int f123() { return 123; }
    virtual int f124() { return 124; }
    virtual int f125() { return 125; }
    virtual int f126() { return 126; }
    virtual int f127() { return 127; }
    virtual ~vptr() = default;
  } _;
  int (vptr::*ptr)() = (int (vptr::*)())f;
  return (_.*ptr)();
}

}  // detail

template <class T>
class GMock {
  static_assert(std::is_polymorphic<T>::value, "T has to be a polymorphic type");
  static_assert(std::has_virtual_destructor<T>::value, "T has to have a virtual destructor");

  unsigned char _[sizeof(T)] = {0};
  void* old_vptr = nullptr;
  void (*vptr[128])() = {
      not_implemented(), not_implemented(), not_implemented(), not_implemented(), not_implemented(), not_implemented(),
      not_implemented(), not_implemented(), not_implemented(), not_implemented(), not_implemented(), not_implemented(),
      not_implemented(), not_implemented(), not_implemented(), not_implemented(), not_implemented(), not_implemented(),
      not_implemented(), not_implemented(), not_implemented(), not_implemented(), not_implemented(), not_implemented(),
      not_implemented(), not_implemented(), not_implemented(), not_implemented(), not_implemented(), not_implemented(),
      not_implemented(), not_implemented(), not_implemented(), not_implemented(), not_implemented(), not_implemented(),
      not_implemented(), not_implemented(), not_implemented(), not_implemented(), not_implemented(), not_implemented(),
      not_implemented(), not_implemented(), not_implemented(), not_implemented(), not_implemented(), not_implemented(),
      not_implemented(), not_implemented(), not_implemented(), not_implemented(), not_implemented(), not_implemented(),
      not_implemented(), not_implemented(), not_implemented(), not_implemented(), not_implemented(), not_implemented(),
      not_implemented(), not_implemented(), not_implemented(), not_implemented(), not_implemented(), not_implemented(),
      not_implemented(), not_implemented(), not_implemented(), not_implemented(), not_implemented(), not_implemented(),
      not_implemented(), not_implemented(), not_implemented(), not_implemented(), not_implemented(), not_implemented(),
      not_implemented(), not_implemented(), not_implemented(), not_implemented(), not_implemented(), not_implemented(),
      not_implemented(), not_implemented(), not_implemented(), not_implemented(), not_implemented(), not_implemented(),
      not_implemented(), not_implemented(), not_implemented(), not_implemented(), not_implemented(), not_implemented(),
      not_implemented(), not_implemented(), not_implemented(), not_implemented(), not_implemented(), not_implemented(),
      not_implemented(), not_implemented(), not_implemented(), not_implemented(), not_implemented(), not_implemented(),
      not_implemented(), not_implemented(), not_implemented(), not_implemented(), not_implemented(), not_implemented(),
      not_implemented(), not_implemented(), not_implemented(), not_implemented(), not_implemented(), not_implemented(),
      not_implemented(), not_implemented(), not_implemented(), not_implemented(), not_implemented(), not_implemented(),
      not_implemented(), not_implemented()};

  auto not_implemented() { return detail::union_cast<void (*)()>(&GMock::unknown); }

  void* unknown() {
    auto* ptr = [this] {
      fs[__PRETTY_FUNCTION__] = std::make_unique<FunctionMocker<void*()>>();
      return static_cast<FunctionMocker<void*()>*>(fs[__PRETTY_FUNCTION__].get());
    }();
    ptr->SetOwnerAndName(this, __PRETTY_FUNCTION__);
    return ptr->Invoke();
  }

  void* update_vptr(void* new_vptr) {
    void* vptr = *(void**)this;
    *(void**)this = new_vptr;
    return vptr;
  }

  template <class TName, class R, class... TArgs>
  decltype(auto) gmock_call_impl(int offset, const detail::identity_t<Matcher<TArgs>>&... args) {
    vptr[offset] = detail::union_cast<void (*)()>(&GMock::template original_call<TName, R, TArgs...>);

    const auto it = fs.find(TName::c_str());
    auto* ptr = [&it, this] {
      if (it == fs.end()) {
        fs[TName::c_str()] = std::make_unique<FunctionMocker<R(TArgs...)>>();
        return static_cast<FunctionMocker<R(TArgs...)>*>(fs[TName::c_str()].get());
      }
      return static_cast<FunctionMocker<R(TArgs...)>*>(it->second.get());
    }();

    ptr->RegisterOwner(this);
    return ptr->With(args...);
  }

  template <class TName, class R, class... TArgs>
  R original_call(TArgs... args) {
    auto* f = static_cast<FunctionMocker<R(TArgs...)>*>(fs[TName::c_str()].get());
    f->SetOwnerAndName(this, TName::c_str());
    return f->Invoke(args...);
  }

 public:
  using type = T;

  GMock() : old_vptr(update_vptr(vptr)) {}

  template <class TName, class R, class B, class... TArgs>
  decltype(auto) gmock_call(R (B::*f)(TArgs...), const detail::identity_t<Matcher<TArgs>>&... args) {
    return gmock_call_impl<TName, R, TArgs...>(detail::vptr_offset(f), args...);
  }

  template <class TName, class R, class B, class... TArgs>
  decltype(auto) gmock_call(R (B::*f)(TArgs...) const, const typename detail::identity_t<Matcher<TArgs>>&... args) {
    return gmock_call_impl<TName, R, TArgs...>(detail::vptr_offset(f), args...);
  }

  T* operator&() { return reinterpret_cast<T*>(this); }
  operator T*() { return reinterpret_cast<T*>(this); }
  operator T&() { return reinterpret_cast<T&>(*this); }

 private:
  std::unordered_map<std::string, std::unique_ptr<internal::UntypedFunctionMockerBase>> fs;
};

template <class T>
using StrictGMock = StrictMock<GMock<T>>;

template <class T>
using NiceGMock = NiceMock<GMock<T>>;

template <class T, class... TArgs>
auto make(TArgs&&... args) {
  return detail::make_impl<T>(/*std::make_tuple(std::forward<TArgs>(args)...)*/);
}

struct uninitialized {};

template <class T>
struct GTest : public Test {
  std::unique_ptr<T> sut;
  std::unordered_map<std::size_t, std::shared_ptr<void>> mocks;

  GTest() { std::tie(sut, mocks) = make<T>(); }

  explicit GTest(uninitialized) {}

  // template <class... TArgs>
  // explicit GTest(TArgs...) {
  // std::tie(sut, mocks) = make<T>(args...);
  //}

  template <class TMock>
  GMock<TMock>& mock() {
    return *static_cast<GMock<TMock>*>(mocks[detail::type_id<TMock>()].get());
  }
};

}  // v1
}  // testing

namespace std {

template <class T, class TDeleter>
auto move(unique_ptr<::testing::GMock<T>, TDeleter>& object) {
  auto* i = object.get();
  return unique_ptr<T>(*i);
}
// static_pointer_cast for shraed_ptr
}

#define __GMOCK_VPTR_COMMA() ,
#define __GMOCK_VPTR_IGNORE(...)
#define __GMOCK_VPTR_NAME(...) __VA_ARGS__ __GMOCK_VPTR_IGNORE
#define __GMOCK_VPTR_QNAME(...) ::testing::detail::string<__GMOCK_VPTR_STR_IMPL_32(#__VA_ARGS__, 0), 0> __GMOCK_VPTR_IGNORE
#define __GMOCK_VPTR_INTERNAL(...)                  \
  __GMOCK_VPTR_IF(__BOOST_DI_IS_EMPTY(__VA_ARGS__)) \
  (__GMOCK_VPTR_IGNORE, __GMOCK_VPTR_COMMA)() __VA_ARGS__
#define __GMOCK_VPTR_CALL(...) __GMOCK_VPTR_INTERNAL
#define __GMOCK_VPTR_PRIMITIVE_CAT(arg, ...) arg##__VA_ARGS__
#define __GMOCK_VPTR_CAT(arg, ...) __GMOCK_VPTR_PRIMITIVE_CAT(arg, __VA_ARGS__)
#define __GMOCK_VPTR_IBP_SPLIT(i, ...) __GMOCK_VPTR_PRIMITIVE_CAT(__GMOCK_VPTR_IBP_SPLIT_, i)(__VA_ARGS__)
#define __GMOCK_VPTR_IBP_SPLIT_0(a, ...) a
#define __GMOCK_VPTR_IBP_SPLIT_1(a, ...) __VA_ARGS__
#define __GMOCK_VPTR_IBP_IS_VARIADIC_C(...) 1
#define __GMOCK_VPTR_IBP_IS_VARIADIC_R_1 1,
#define __GMOCK_VPTR_IBP_IS_VARIADIC_R___GMOCK_VPTR_IBP_IS_VARIADIC_C 0,
#define __GMOCK_VPTR_IBP(...) \
  __GMOCK_VPTR_IBP_SPLIT(0, __GMOCK_VPTR_CAT(__GMOCK_VPTR_IBP_IS_VARIADIC_R_, __GMOCK_VPTR_IBP_IS_VARIADIC_C __VA_ARGS__))
#define __BOOST_DI_IS_EMPTY(...)                           \
  __GMOCK_VPTR_IS_EMPTY_IIF(__GMOCK_VPTR_IBP(__VA_ARGS__)) \
  (__GMOCK_VPTR_IS_EMPTY_GEN_ZERO, __GMOCK_VPTR_IS_EMPTY_PROCESS)(__VA_ARGS__)
#define __GMOCK_VPTR_IS_EMPTY_PRIMITIVE_CAT(a, b) a##b
#define __GMOCK_VPTR_IS_EMPTY_IIF(bit) __GMOCK_VPTR_IS_EMPTY_PRIMITIVE_CAT(__GMOCK_VPTR_IS_EMPTY_IIF_, bit)
#define __GMOCK_VPTR_IS_EMPTY_NON_FUNCTION_C(...) ()
#define __GMOCK_VPTR_IS_EMPTY_GEN_ZERO(...) 0
#define __GMOCK_VPTR_IS_EMPTY_IIF_0(t, b) b
#define __GMOCK_VPTR_IS_EMPTY_IIF_1(t, b) t
#define __GMOCK_VPTR_IS_EMPTY_PROCESS(...) __GMOCK_VPTR_IBP(__GMOCK_VPTR_IS_EMPTY_NON_FUNCTION_C __VA_ARGS__())
#define __GMOCK_VPTR_IIF(c) __GMOCK_VPTR_PRIMITIVE_CAT(__GMOCK_VPTR_IIF_, c)
#define __GMOCK_VPTR_IIF_0(t, ...) __VA_ARGS__
#define __GMOCK_VPTR_IIF_1(t, ...) t
#define __GMOCK_VPTR_IF(c) __GMOCK_VPTR_IIF(c)
#define __GMOCK_VPTR_STR_IMPL_1(str, i) (sizeof(str) > (i) ? str[(i)] : 0)
#define __GMOCK_VPTR_STR_IMPL_4(str, i) \
  __GMOCK_VPTR_STR_IMPL_1(str, i + 0)   \
  , __GMOCK_VPTR_STR_IMPL_1(str, i + 1), __GMOCK_VPTR_STR_IMPL_1(str, i + 2), __GMOCK_VPTR_STR_IMPL_1(str, i + 3)
#define __GMOCK_VPTR_STR_IMPL_16(str, i) \
  __GMOCK_VPTR_STR_IMPL_4(str, i + 0)    \
  , __GMOCK_VPTR_STR_IMPL_4(str, i + 4), __GMOCK_VPTR_STR_IMPL_4(str, i + 8), __GMOCK_VPTR_STR_IMPL_4(str, i + 12)
#define __GMOCK_VPTR_STR_IMPL_32(str, i) \
  __GMOCK_VPTR_STR_IMPL_16(str, i + 0)   \
  , __GMOCK_VPTR_STR_IMPL_16(str, i + 16), __GMOCK_VPTR_STR_IMPL_16(str, i + 32)

#undef EXPECT_CALL
#define __GMOCK_VPTR_EXPECT_CALL_0(obj, call) GMOCK_EXPECT_CALL_IMPL_(obj, call)
#define __GMOCK_VPTR_EXPECT_CALL_1(obj, call)                                              \
  ((obj).gmock_call<__GMOCK_VPTR_QNAME call>(                                              \
       &std::decay_t<decltype(obj)>::type::__GMOCK_VPTR_NAME call __GMOCK_VPTR_CALL call)) \
      .InternalExpectedAt(__FILE__, __LINE__, #obj, #call)
#define EXPECT_CALL(obj, call) __GMOCK_VPTR_CAT(__GMOCK_VPTR_EXPECT_CALL_, __GMOCK_VPTR_IBP(call))(obj, call)

#undef ON_CALL
#define __GMOCK_VPTR_ON_CALL_0(obj, call) GMOCK_ON_CALL_IMPL_(obj, call)
#define __GMOCK_VPTR_ON_CALL_1(obj, call)                                                  \
  ((obj).gmock_call<__GMOCK_VPTR_QNAME call>(                                              \
       &std::decay_t<decltype(obj)>::type::__GMOCK_VPTR_NAME call __GMOCK_VPTR_CALL call)) \
      .InternalDefaultActionSetAt(__FILE__, __LINE__, #obj, #call)
#define ON_CALL(obj, call) __GMOCK_VPTR_CAT(__GMOCK_VPTR_ON_CALL_, __GMOCK_VPTR_IBP(call))(obj, call)
